files-community/Files
# Code Quality: Disable drag & drop when running as admin

### Summary

The service 'IWindowsSecurityService' gets the current process's token and checks if the token belongs to Administrators group.

- [x] Create `IWindowsSecurityService`
- [x] Replace ElevationHelpers with the service above
- [x] Cache the result in `WindowContext`

### Resolved / Related Issues

- Closes #13394

### Steps used to test these changes

1. Open Files app as admin
2. Confirm drag and drop is disabled in TabBar and content views.
3. Check it also when UAC is disabled.

<details><summary>Open details that I learnt</summary><div>

### Why drag and drop doesn’t work?

This is because of UIPI (User Interface Privilege Isolation) using MIC (Mandatory Integrity Control), which blocks a process that has lower IL (integrity level) from interacting with a process that has higher IL. This was introduced in Vista with UAC introduction.

It is not caused by UIPI, btw, that UWP cannot access system resource and user data without user’s consent (it’s by AppContainer itself).

IL get higher from top to bottom

1. `Untrusted` (Chrome, IE)
2. `AppContainer/LowBox` (UWP)
4. `Low`
5. `Medium` (default)
6. `Medium Plus`
7. `High/Elevated` (elevated)
8. `System` (system services)
9. `Protected Process`
10. `Installer` (Windows)

### UAC settings

**Behavior**

Level|Modification<br/>by a process|Modification<br/>by user|On what session|Remarks
:-:|---|---|---|---
Highest|Yes|Yes|Secure Desktop (dimmed)|Default in Vista
2nd|Yes|No|Secure Desktop (dimmed)|Default in 7 onwards
3rd|Yes|No|Normal Desktop (not dimmed)|Not recommended
Lowest|No|No|N/A|Not recommended

- By process means: Installation of a software manifest of which calls for elevation, and invocation via RunAs verb
- By user means: modification via Windows Settings app and certain Control Panel applets


**Registry values to be modified**

Level|ConsentPrompt<br/>BehaviorAdmin|ConsentPrompt<br/>BehaviorUser|EnableLUA|PromptOn<br/>SecureDesktop
:-:|---|---|---|---
Highest|2 (AAC UAC)|3 (OTS UAC)|1|1
2nd|5|3|1|1
3rd|5|3|1|0
Lowest|0|3|0|0

- AAC (Admin Approval Consent) UAC displays only Yes or No since logon user is in Administrators group.
- OTS (Over The Shoulder) UAC displays text boxes to enter credential of an admin identity.

### Disabling UAC

When you disabled UAC through UAC Settings dialog (UIPI will be disabled as well), the UAC prompt won’t be shown.

However, it seems that DataExchangeHost.exe doesn’t accept dropping onto a window of a process that has higher IL (Windows OS Bug?) even though UAC is disabled. A contributor of Windows Terminal made a workaround for it.
Also, I should denote that we might be able to use DragQueryFile to workaround UIPI, while I’m still not sure.

### How UAC works (if it helps)?

On startup, the kernel executes explorer.exe with a token below:

**Standard User (not in Administrators group)**

- Only *restricted* token will be generated after logon.
- ShellExecute calls appinfo.exe (AIS) and then it calls consent.exe (UAC prompt) via its hosted inner service. When a valid credential of an administrator entered, AIS calls CreateProcessAsUser with that administrative identity; otherwise, it returns ERROR_ACCESS_DENIED.
- The token generated by this routine is called *filtered* token (not *privileged* token)

**Administrator User (in Administrators group)**

- Both *privileged* and *restricted* token will be generated.
- When Yes clicked, AIS calls CreateProcessAsUser with the logon user ‘s *privileged* token; otherwise, it returns ERROR_ACCESS_DENIED.

**BUILTIN/Administrator (the real administrator)**

- Only *privileged* token will be generated
- AIS doesn’t call consent.exe. But this behavior can be altered from group policy.

---

I've been described so far, for when user hasn't changed anything through Windows Registry or Group Policy. Some behavior can be changed otherwise.

</div></details>
